File: src/ClaudeUsageWatcher.Core/BurnRateEstimator.cs
namespace ClaudeUsageWatcher.Core;

public sealed class BurnRateEstimator
{
    private const int Capacity = 10;
    private const double DropThreshold = 5.0;
    private const double MinSpanMinutes = 3.0;
    private const double MinRate = 0.05;

    private readonly (DateTimeOffset At, double Utilization)[] _buffer =
        new (DateTimeOffset, double)[Capacity];
    private int _count;
    private int _head; // index of the oldest sample

    public void Add(DateTimeOffset at, double utilization)
    {
        if (_count > 0)
        {
            int lastIndex = (_head + _count - 1) % Capacity;
            double lastUtil = _buffer[lastIndex].Utilization;
            if (utilization < lastUtil - DropThreshold)
            {
                // Window reset detected: discard history and restart.
                _count = 0;
                _head = 0;
            }
        }

        if (_count < Capacity)
        {
            int writeIndex = (_head + _count) % Capacity;
            _buffer[writeIndex] = (at, utilization);
            _count++;
        }
        else
        {
            _buffer[_head] = (at, utilization);
            _head = (_head + 1) % Capacity;
        }
    }

    public TimeSpan? ProjectTimeToCap(DateTimeOffset now)
    {
        if (_count < 2)
        {
            return null;
        }

        (DateTimeOffset At, double Utilization) first = _buffer[_head];
        (DateTimeOffset At, double Utilization) last = _buffer[(_head + _count - 1) % Capacity];

        double spanMinutes = (last.At - first.At).TotalMinutes;
        if (spanMinutes < MinSpanMinutes)
        {
            return null;
        }

        double rate = (last.Utilization - first.Utilization) / spanMinutes;
        if (rate <= MinRate)
        {
            return null;
        }

        double minutesToCap = (100 - last.Utilization) / rate;
        DateTimeOffset capInstant = last.At.AddMinutes(minutesToCap);
        TimeSpan remaining = capInstant - now;
        return remaining < TimeSpan.Zero ? TimeSpan.Zero : remaining;
    }
}

File: macos/Sources/UsageCore/BurnRateEstimator.swift
import Foundation

/// Ring-buffer burn-rate estimator. Capacity 10; a sample that drops more than
/// 5 percentage points below the previous one is treated as a window reset and
/// clears the history (the usage window rolled over, so prior slope is moot).
public final class BurnRateEstimator {
    private static let capacity = 10
    private static let dropThreshold = 5.0
    private static let minSpanMinutes = 3.0
    private static let minRate = 0.05

    private var buffer: [(at: Date, utilization: Double)] = []
    private var count = 0
    private var head = 0

    public init() {
        buffer = Array(repeating: (Date(timeIntervalSince1970: 0), 0.0), count: Self.capacity)
    }

    public func add(at: Date, utilization: Double) {
        if count > 0 {
            let lastIndex = (head + count - 1) % Self.capacity
            let lastUtil = buffer[lastIndex].utilization
            if utilization < lastUtil - Self.dropThreshold {
                // Window reset detected: discard history and restart.
                count = 0
                head = 0
            }
        }

        if count < Self.capacity {
            let writeIndex = (head + count) % Self.capacity
            buffer[writeIndex] = (at, utilization)
            count += 1
        } else {
            buffer[head] = (at, utilization)
            head = (head + 1) % Self.capacity
        }
    }

    /// Returns time remaining until the window hits 100%, or nil when the slope
    /// is too shallow / sparse to be meaningful. Clamped at zero.
    public func projectTimeToCap(now: Date) -> TimeInterval? {
        if count < 2 {
            return nil
        }

        let first = buffer[head]
        let last = buffer[(head + count - 1) % Self.capacity]

        let spanMinutes = last.at.timeIntervalSince(first.at) / 60.0
        if spanMinutes < Self.minSpanMinutes {
            return nil
        }

        let rate = (last.utilization - first.utilization) / spanMinutes
        if rate <= Self.minRate {
            return nil
        }

        let minutesToCap = (100 - last.utilization) / rate
        let capInstant = last.at.addingTimeInterval(minutesToCap * 60.0)
        let remaining = capInstant.timeIntervalSince(now)
        return remaining < 0 ? 0 : remaining
    }
}
